File: src/service/userService.ts
import jwt from "jsonwebtoken";
import md5 from "md5";
import bcrypt from "bcryptjs";
import crypto from "crypto";
import sgMail from '@sendgrid/mail';

import {User} from "../models/User";
import { AuthResponse, JwtPayload, RegisterResponse, ForgotPasswordResponse, ResetPasswordResponse } from "types";


const config = require('config.json')


export async function Authenticate(username: string, password: string): Promise<AuthResponse> {
    return User.scope('auth').findOne({where: {
        username: username.toLowerCase()
    }}).then(user => {
        if (user) {
            if (user.hash.startsWith("old:") && (md5(password.trim()) === user.hash.trim().substr(4))) {
                const token = jwt.sign({ id: user.id, level: user.userlevel, passwordreset: user.passwordreset} as JwtPayload, config.secret, {expiresIn: 24*60*60});
                return { success: true, message: 'Authentication Successful', token: token};
            } else if (bcrypt.compareSync(password, user.hash)) {
                const token = jwt.sign({ id: user.id, level: user.userlevel, passwordreset: user.passwordreset} as JwtPayload, config.secret, {expiresIn: 24*60*60});
                return { success: true, message: 'Authentication Successful', token: token};
            } else {
                console.debug('User failed auth: %s', username);
                return { success: false, message: 'Authentication Failed: User or Password invalid.', token: ''};
            }
            
        } else {
            console.debug('Failed to find user: %s', username);
            return { success: false, message: 'Authentication Failed: User or Password invalid.', token: ''};
        }
    }).catch(err => {
        console.warn(err.message);
        return { success: false, message: 'Authentication Failed: User or Password invalid.', token: ''};
    });
}

export async function Register(username: string, password: string, email: string,
                            firstname: string, lastname: string): Promise<RegisterResponse> {
    const usernameLower = username.toLowerCase();
    const emailLower = email.toLowerCase();
    const failedMessage: RegisterResponse = {success: false, message: "Failed to register user."};
    // Check username and email not used
    const usernameUsed = await usernameTaken(usernameLower).catch(err => {return failedMessage});
    if (usernameUsed) {
        if (typeof usernameUsed === 'boolean') {
            return {success: false, message: "Username Already used!"};
        } else {
            return usernameUsed as unknown as RegisterResponse;
        }
    }
    const emailUsed = await emailTaken(emailLower).catch(err => {return failedMessage});
    if (emailUsed) {
        if (typeof emailUsed === 'boolean') {
            return {success: false, message: "Email Already used!"};
        } else {
        return emailUsed as unknown as RegisterResponse;
        }
    }

    const hash = await hashPassword(password);
    password = "";

    return User.create({username: usernameLower, hash: hash, email: emailLower, firstname: firstname, lastname: lastname, userlevel: 1})
        .then(user=> {
            if (user) {
                return {success: true, message: "User created."}
            } else {
                return {success: false, message: "User NOT created."}
            }
        }).catch(err => {
            console.error(err);
            return {success: false, message: "User NOT created."}
        });
}

export async function PasswordForgot(email: string, hostUrl: string): Promise<ForgotPasswordResponse> {
    return User.findOne({where: {email: email.toLowerCase()}}).then(user => {
        if (user) {
            const token = crypto.randomBytes(20).toString('hex');
            user.passwordResetToken = token;
            user.passwordResetExpires = new Date(Date.now()+3600000);
            return user.save().then(user => {
                //send email
                sgMail.setApiKey(config.sendgridAPIKey);
                var message = {
                    to: user.email,
                    from: 'noreply@jatgam.com',
                    subject: 'Wishlist Password Reset',
                    text: 'You are receiving this because you (or someone else) requested the reset of the password for your account.\n\n'+
                        'Please click the following link, or paste into your browser to complete the process:\n\n'+
                        'https://'+hostUrl+'/password_reset/'+token+'\n\n'+
                        'Username: '+user.username+'\n\n'+
                        'If you did not request this, please ignore this email and your password will remain unchanged.'
                };
                return sgMail.send(message).then(() => {
                    console.log('Sending Password Reset Email to: '+user.email);
                    return {success: true, message: 'Sending an Email to the provided address.'};
                }).catch(error => {
                    console.error(error.toString());
                    return {success: false, message: 'Error Sending Password Reset Email'};
                });
            }).catch(err => {
                console.error(err);
                return {success: false, message: 'Error Trying to Reset Password'};
            });
        } else {
            return {success: false, message: 'No account with that Email.'};
        }
    });
}

export async function PasswordReset(email: string, password: string, token: string): Promise<ResetPasswordResponse> {
    if (await PasswordResetTokenValid(token)) {
        return User.scope('auth').findOne({where: {email: email.toLowerCase(), passwordResetToken: token}})
            .then(async user => {
                if (user) {
                    user.hash = await hashPassword(password);
                    user.passwordreset = false;
                    user.passwordResetToken = null;
                    user.passwordResetExpires = null;

                    return user.save().then(user => {
                        console.debug("Reset password for: %s", user.username);
                        return {success: true, message: 'Password Reset'};
                    }).catch(err => {
                        console.error(err);
                        return {success: false, message: 'Failed to rest the password'};
                    });
                } else {
                    return {success: false, message: 'Password reset token is invalid or expired.'};
                }
            });
    } else {
        return {success: false, message: 'Password reset token is invalid or expired.'};
    }
}

export async function PasswordResetTokenValid(token: string): Promise<boolean> {
    return User.scope('passReset').findOne({where: {passwordResetToken: token}}).then(user => {
        if (user && user.passwordResetToken && user.passwordResetExpires) {
            if (user.passwordResetToken.length < 40) {
                return false;
            }
            if ((token === user.passwordResetToken) && (user.passwordResetExpires.getTime() > new Date(Date.now()).getTime())) {
                return true;
            } else {
                return false;
            }
        } else {
            return false;
        }
    }).catch(err => {
        console.error(err);
        return false;
    });
}

export async function getById(id: number) {
    return await User.findByPk(id);
}

async function usernameTaken(username: string): Promise<boolean> {
    return User.findOne({where: {username: username.toLowerCase()}}).then(user => {
        if (user) {
            return true;
        } else {
            return false;
        }
    }).catch(err => {
        console.error(err.message);
        throw new Error(err)
    })
}

async function emailTaken(email: string): Promise<boolean> {
    return User.findOne({where: {email: email.toLowerCase()}}).then(email => {
        if (email) {
            return true;
        } else {
            return false;
        }
    }).catch(err => {
        console.error(err.message);
        throw new Error(err);
    })
}

async function hashPassword(password: string): Promise<string> {
    return bcrypt.hash(password, 13);
}

File: service/userService.go
package service

import (
	"fmt"
	"strings"
	"time"

	"github.com/sirupsen/logrus"

	"github.com/jatgam/wishlist-api/models"
	"github.com/jatgam/wishlist-api/service/sgmail"
	"github.com/jatgam/wishlist-api/types"
	"github.com/jatgam/wishlist-api/utils"
)

func RegisterUser(username, password, email, firstname, lastname string, logger *logrus.Entry) error {
	username = strings.ToLower(username)
	email = strings.ToLower(email)
	userTaken, err := usernameTaken(username)
	if err != nil {
		logger.Errorf("User Registration Failed: %s", err)
		return types.ErrUserRegister
	}
	if userTaken {
		logger.Debugf("Username Taken: %s", username)
		return types.ErrUsernameTaken
	}
	emTaken, err := emailTaken(email)
	if err != nil {
		logger.Errorf("User Registration Failed: %s", err)
		return types.ErrUserRegister
	}
	if emTaken {
		logger.Debugf("EMail Taken: %s", email)
		return types.ErrEmailTaken
	}
	hash, err := utils.HashPassword(password)
	if err != nil {
		logger.Errorf("User Registration Password Hash Failed: %s", err)
		return types.ErrUserRegister
	}
	newUser := &models.UserModel{Username: username, PasswordHash: hash,
		EMail: email, FirstName: firstname, LastName: lastname, UserLevel: 1}
	err = models.CreateUser(newUser)
	if err != nil {
		logger.Errorf("User Registration DB Insert Failed: %s", err)
		return types.ErrUserRegister
	}
	return nil
}

func PasswordForgot(email, hostUrl string, logger *logrus.Entry) error {
	email = strings.ToLower(email)
	user, err := models.FindOneUser(&models.UserModel{EMail: email}, models.UserPassResetScope)
	if err != nil {
		logger.Errorf("PasswordForgot: Failed DB query: %s", email)
		return types.ErrPasswordForgot
	}
	if user == nil {
		logger.Errorf("PasswordForgot: No user record: %s", email)
		return types.ErrPasswordForgot
	}

	token, err := utils.RandomHex(20)
	if err != nil {
		logger.Errorf("PasswordForgot: Failed to Generate Reset Token: %s", err.Error())
		return types.ErrPasswordForgot
	}

	expirationTime := time.Now().Add(time.Hour)
	updates := models.UserModel{PasswordResetToken: &token, PasswordResetExpires: &expirationTime}
	err = models.UpdateUser(user, updates)
	if err != nil {
		logger.Errorf("PasswordForgot: Failed to Update User in DB: %s", err.Error())
		return types.ErrPasswordForgot
	}

	message := fmt.Sprintf("You are receiving this because you (or someone else) requested the reset of the password for your account.\n\n"+
		"Please click the following link, or paste into your browser to complete the process:\n\n"+
		"https://%s/password_reset/%s\n\n"+
		"Username: %s\n\n"+
		"If you did not request this, please ignore this email and your password will remain unchanged.",
		hostUrl, token, user.Username)

	mailer := sgmail.GetMailer()
	err = mailer.SendMail(email, "Jatgam Wishlist Password Reset", message, logger)
	if err != nil {
		logger.Errorf("PasswordForgot: Error Sending Reset Email: %s", err.Error())
		return types.ErrPasswordForgot
	}
	return nil
}

func PasswordResetTokenValidate(token string, logger *logrus.Entry) (bool, error) {
	user, err := models.FindOneUser(&models.UserModel{PasswordResetToken: &token}, models.UserPassResetScope)
	if err != nil {
		logger.Errorf("PasswordResetTokenValidate: Failed DB query: %s", token)
		return false, types.ErrPasswordResetValidate
	}
	if user == nil {
		logger.Errorf("PasswordResetTokenValidate: No user record: %s", token)
		return false, types.ErrPasswordResetValidate
	}

	if user.PasswordResetToken != nil && user.PasswordResetExpires != nil {
		if len(*user.PasswordResetToken) == 40 && *user.PasswordResetToken == token && user.PasswordResetExpires.After(time.Now()) {
			logger.Debugf("PasswordReset Token is Valid: %s", token)
			return true, nil
		}
	}
	logger.Debugf("PasswordReset Token is INVALID: %s", token)
	return false, nil
}

func PasswordReset(email, password, token string, logger *logrus.Entry) error {
	email = strings.ToLower(email)
	tokenValid, _ := PasswordResetTokenValidate(token, logger)
	if !tokenValid {
		logger.Debugf("Password Reset for %s:%s had invalid token.", email, token)
		return types.ErrPasswordResetValidate
	}

	user, err := models.FindOneUser(&models.UserModel{EMail: email}, models.UserAuthScope)
	if err != nil {
		logger.Errorf("PasswordReset: Failed DB query: %s", email)
		return types.ErrPasswordReset
	}
	if user == nil {
		logger.Errorf("PasswordReset: No user record: %s", email)
		return types.ErrPasswordReset
	}

	if user.PasswordResetToken != nil && *user.PasswordResetToken != token {
		logger.Errorf("PasswordReset: Potential multiple of same reset tokens exist")
		return types.ErrPasswordResetValidate
	}

	hash, err := utils.HashPassword(password)
	if err != nil {
		logger.Errorf("PasswordReset Password Hash Failed: %s", err)
		return types.ErrPasswordReset
	}
	updates := map[string]interface{}{"PasswordHash": hash, "PasswordReset": false, "PasswordResetToken": nil, "PasswordResetExpires": nil}
	// updates := models.UserModel{PasswordHash: hash, PasswordReset: false, PasswordResetToken: nil, PasswordResetExpires: nil}
	err = models.UpdateUserWithMap(user, updates)
	if err != nil {
		logger.Errorf("PasswordReset: Failed to Update User in DB: %s", err.Error())
		return types.ErrPasswordReset
	}

	return nil
}

func usernameTaken(username string) (bool, error) {
	user, err := models.FindOneUser(&models.UserModel{Username: username})
	if err != nil {
		return true, err
	}
	if user != nil {
		return true, nil
	}
	return false, nil
}

func emailTaken(email string) (bool, error) {
	user, err := models.FindOneUser(&models.UserModel{EMail: email})
	if err != nil {
		return true, err
	}
	if user != nil {
		return true, nil
	}
	return false, nil
}
